### PR TITLE
Issue/2711 Add downloadable files to a product

### LINKED ISSUE #2711: Add/Edit Products M5: Downloadable files Step 4: add a downloadable file for a product
As part of the M5 release, we need to provide the option to add a new downloadable product by:
- choosing a file from the mobile device and uploading the file.
- choosing a file from the media picker library.
- manually entering a file url, the download expiry and download limit.

This issue should also take care to display the downloadable files as part of the bottom sheet, if there are no existing files for the product.
Screenshot 2020-08-09 at 12 56 15 AM

#### Choose to add a downloadable file
Screenshot 2020-08-09 at 12 55 47 AM

Clicking on the Add file option should display a bottom sheet with the following options:
Screenshot 2020-08-09 at 12 58 34 AM

#### Choose from device
Screenshot 2020-08-09 at 12 59 26 AM

#### Choose from media picker library
Screenshot 2020-08-09 at 12 59 35 AM

#### Manually add a file
Screenshot 2020-08-09 at 12 59 42 AM

Feel free to split this issue into smaller parts!

### PR DESCRIPTION
**Pull request by hichamboushaba**
Fixes #2711 

**Description**
This PR add the implementation of adding downloadable files flow to products, the list of important changes:

- Adds an entry to the "Add more details" sheet to allow adding downloadable files if the product wasn't downloadable, or didn't have any downloadable files before.
- Created [MediaFilesRepository](https://github.com/woocommerce/woocommerce-android/blob/8ee07126e286d8d8d3c7498bbcd1347462e449f6/WooCommerce/src/main/kotlin/com/woocommerce/android/media/MediaFilesRepository.kt) to expose a coroutine for uploading files.
- Modified the viewstates to allow displaying a progress dialog when a file is being uploaded.
- Made some changes to the WPMediaGalleryView to allow single select mode.
- Made some changes to WPMediaPickerFragment to allow defining requestCode in navArgs, and use it for the returned result.
- https://github.com/woocommerce/woocommerce-android/pull/2760/files#r474648485
- Added input validation for the url/name in the file details.

**Screenshots**
Screen Shot 2020-08-21 at 23 14 20  Screen Shot 2020-08-24 at 11 31 38
Screen Shot 2020-08-21 at 23 15 05 Screen Shot 2020-08-21 at 23 15 14
Screen Shot 2020-08-24 at 11 31 47 Screen Shot 2020-08-21 at 23 17 07

**Testing**

**Test 1: Product not downloadable (or downloadable files is empty)**
1. Open product details.
2. Click on "Add more details"
3. Select "Downloadable files"
4. Make sure that a bottom sheet is displayed listing different sources to pick a file.
5. Test different scenarios, and confirm that when picking a file from device or camera, a progress dialog is shown while uploading the file.
6. Edit details of the file in the details screen, and save.
7. Confirm that the file was added.

**Test 2: Product has downloadable files**
1. Open product details.
2. Click on "downloadable files".
3. Click on "Add file"
4. same as steps 4-7 from test 1.
 
**Test 3: Input validation**
Make sure that input validation respects the following cases:
- URL empty => show the error
- URL non empty but doesn’t contain a path (something like: http://testurl.com/):
           - Name is empty => show an error that the name is missing.
           - Name non empty => input is valid, enable the done button.
- URL non empty and have a path (http://testurl.com/file.jpg) => input is valid regardless of the content of the name field.
- URL has to be a valid URL, if the user doesn’t enter a protocol, we’ll add “http” implicitly.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/wpmediapicker/WPMediaPickerFragment.kt [hunk 1]
```diff
@@ -137,10 +140,10 @@ class WPMediaPickerFragment : BaseFragment(), WPMediaGalleryListener, BackPressL
...
             }
             requireActivity().navigateBackWithResult(
-                    RequestCodes.WPMEDIA_LIBRARY_PICKER,
-                    bundle,
-                    R.id.nav_host_fragment_main,
-                    R.id.productDetailFragment
+                navArgs.requestCode,
+                bundle,
+                R.id.nav_host_fragment_main,
+                findNavController().previousBackStackEntry?.destination?.id ?: R.id.productDetailFragment
```
Comment (hichamboushaba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/wpmediapicker/WPMediaPickerFragment.kt:146]: Having the R.id.productDetailFragment static here meant that we always popstack all the way to the ProductDetailFragment, which causes an issue after adding a file from the ProductDownloadsFragment.
I updated it to popstack to just the previous fragment, and updated ProductImagesFragment to intercept the result of picking images.
And this improved also the UX as the flow is similar to picking from camera/device now, check below gifs for the new behavior:

Before:
wpmedia_pick_file_before

After:
wpmedia_pick_file_after

Comment (anitaa1990) [reply WooCommerce/src/main/kotlin/com/woocommerce/android/ui/wpmediapicker/WPMediaPickerFragment.kt:146]: Nice <3

**Review by anitaa1990 — CHANGES_REQUESTED** [review]
Nice work on this @hichamboushaba!! I just have a few comments after testing:

- I have a product with the downloadable flag switched on and the product does not have any downloadable files so it should be displayed in the bottom sheet right? I am not seeing that though.
- When choosing files from a device, we have the option to select multiple files but we only want to support single selection right?
- When app is resumed from background and `do not keep activities` is enabled, the added file is still displayed but the “Update” menu button is not, so there is no way to update. I’m guessing this has something to do with the fix you made on `develop`? If so, we can leave it as is :) 
- Can we update the progress dialog message when uploading a file? It says “Loading”, which feels incorrect. Maybe uploading file? 
- I see from the designs that we also have the option to manually enter file url and name in thee bottom sheet. Is that being done in another PR?

Screenshot 2020-08-24 at 10 42 53 AM

- When clearing the contents of the url + name of an exiting file, the update menu button is disabled but there is no error message displayed.


- When the name is empty and the url does not have a path, I am not seeing the error message and the UPDATE menu button is displayed, when it should be disabled?

**Comment by hichamboushaba** [comment]
> * I have a product with the downloadable flag switched on and the product does not have any downloadable files so it should be displayed in the bottom sheet right? I am not seeing that though.

On my side I see the item in the bottom sheet, can you please send me the returned payload from the backend to try to understand why it's happening for you.

> * When choosing files from a device, we have the option to select multiple files but we only want to support single selection right?

You are right, it hould be fixed now.

> * When app is resumed from background and `do not keep activities` is enabled, the added file is still displayed but the “Update” menu button is not, so there is no way to update. I’m guessing this has something to do with the fix you made on `develop`? If so, we can leave it as is :)

On my side I saw a different behavior, the file was removed, and after checking I found that I missed updating the `mergeProduct` function, it's done on 7ba7c277655f2d9f806beb87504320223225726c.

> * Can we update the progress dialog message when uploading a file? It says “Loading”, which feels incorrect. Maybe uploading file?

@Garance91540 what do you think?

> * I see from the designs that we also have the option to manually enter file url and name in thee bottom sheet. Is that being done in another PR?

I totally missed this part, I thought it had the same design as the add photos dialog 🤦 .
I updated it, and removed the class FilePickerChooserView as it doesn't make sense now.

> * When clearing the contents of the url + name of an exiting file, the update menu button is disabled but there is no error message displayed.

The issue is that the `WCMaterialOutlinedEditTextView` class was clearing the error automatically every time the text is changed, and as we re-apply the error only if it changed, we weren't reapplying it.
I made some updates on the commit: c3c7f4026d5c9f447fe93af0e987419f8b3e4173 to avoid clearing the error automatically, my understanding is that this was needed for `onSkuChanged`, so i updated its logic to clear the error from the viewmodel. 

> * When the name is empty and the url does not have a path, I am not seeing the error message and the UPDATE menu button is displayed, when it should be disabled?

You are right, I updated the validation rules: c7d34a77f7f9e147496d590511c755893041a7c3

**Comment by Garance91540** [comment]
@hichamboushaba , thank you for working on this! Is it possible to send me some screenshots for Dark mode + tablet too?
A light mode, phone screencast would also be helpful to get a better understanding of the flow

> Can we update the progress dialog message when uploading a file? It says “Loading”, which feels incorrect. Maybe uploading file?
@Garance91540 what do you think?

"Uploading files" sounds good 👍

**Comment by hichamboushaba** [comment]
> thank you for working on this! Is it possible to send me some screenshots for Dark mode + tablet too?
A light mode, phone screencast would also be helpful to get a better understanding of the flow

@Garance91540 please find below some screenshots, if you feel something is missing, let me know and I'll add it.

**Dark mode**
Screen Shot 2020-08-24 at 16 26 41 Screen Shot 2020-08-24 at 16 26 46
Screen Shot 2020-08-24 at 16 33 05 Screen Shot 2020-08-24 at 16 33 41

**Tablet**
Screen Shot 2020-08-24 at 16 40 30
Screen Shot 2020-08-24 at 16 41 04
Screen Shot 2020-08-24 at 16 41 17

**Screencast**
add_files

**Comment by Garance91540** [comment]
@hichamboushaba this is great, thank you for working on this!

**Review by anitaa1990 — CHANGES_REQUESTED** [review]
> I have a product with the downloadable flag switched on and the product does not have any downloadable files so it should be displayed in the bottom sheet right? I am not seeing that though.
On my side I see the item in the bottom sheet, can you please send me the returned payload from the backend to try to understand why it's happening for you.

Ahh, I found out the confusion here. It looks like the `downloadable` flag should only be displayed for Simple products. Th product I mentioned yesterday was an external product but the `downloadable` flag was displayed so I switched it on. But I didn’t see the `Downloadable` files section in the bottom sheet. Can we make sure to display the flag only if thee product is `Simple`?

- When clicking on manually enter url, I noticed that the error message is displayed even when the file url has a path and a file name. See below screenshot:



- When validating a file, I understand that when there is no path to a file url and the file name is empty, we should display an error. But I noticed that when the file url has no path and I remove the name of the file, the error displayed is something like “Enter a valid file url”, which is really confusing. @garance can confirm here but it would make more sense to display the error in this case, on the file name, instead of the file url, since I made changes to the file name only and not the url.

**Comment by hichamboushaba** [comment]
> Ahh, I found out the confusion here. It looks like the `downloadable` flag should only be displayed for Simple products. Th product I mentioned yesterday was an external product but the `downloadable` flag was displayed so I switched it on. But I didn’t see the `Downloadable` files section in the bottom sheet. Can we make sure to display the flag only if thee product is `Simple`?

Sure, I'll do it.

> * When clicking on manually enter url, I noticed that the error message is displayed even when the file url has a path and a file name. See below screenshot:

You have to specify a scheme for your url (http:// or https://), because otherwise the backend considers that the url is a relative path in the server, and returns an error that it doesn't exist:

```json
{
    "code": "product_invalid_download",
    "message": "The downloadable file <code>test.com/file.jpg</code> cannot be used as it does not exist on the server.",
    "data": {
        "status": 400
    }
}
```

> * When validating a file, I understand that when there is no path to a file url and the file name is empty, we should display an error. But I noticed that when the file url has no path and I remove the name of the file, the error displayed is something like “Enter a valid file url”, which is really confusing. @garance can confirm here but it would make more sense to display the error in this case, on the file name, instead of the file url, since I made changes to the file name only and not the url.

Yes, I asked about this in my p2 comment, but when I got the answer, it confirmed the initial validation 🙂 , so I assumed it's what we want.

**Review by anitaa1990 — COMMENTED** [review]
> Ahh, I found out the confusion here. It looks like the downloadable flag should only be displayed for Simple products. Th product I mentioned yesterday was an external product but the downloadable flag was displayed so I switched it on. But I didn’t see the Downloadable files section in the bottom sheet. Can we make sure to display the flag only if thee product is Simple?

Works great now 🎉 

> When clicking on manually enter url, I noticed that the error message is displayed even when the file url has a path and a file name. See below screenshot:
You have to specify a scheme for your url (http:// or https://), because otherwise the backend considers that the url is a relative path in the server, and returns an error that it doesn't exist:

Ah, fair enough. But can we make sure to add the scheme implicitly? I seem to remember we do a similar logic during login. It's not likely that users will type the scheme when entering a url so it makes sense to add it ourselves. WDYT?

> When validating a file, I understand that when there is no path to a file url and the file name is empty, we should display an error. But I noticed that when the file url has no path and I remove the name of the file, the error displayed is something like “Enter a valid file url”, which is really confusing. @garance can confirm here but it would make more sense to display the error in this case, on the file name, instead of the file url, since I made changes to the file name only and not the url.
Yes, I asked about this in my p2 comment, but when I got the answer, it confirmed the initial validation 🙂 , so I assumed it's what we want.

Based on the discussion from yesterday, we are going to be adding the error to the filename. Is that correct?

**Comment by hichamboushaba** [comment]
> Ah, fair enough. But can we make sure to add the scheme implicitly? I seem to remember we do a similar logic during login. It's not likely that users will type the scheme when entering a url so it makes sense to add it ourselves. WDYT?

> Based on the discussion from yesterday, we are going to be adding the error to the filename. Is that correct?

Both are done, please check the commit 5fbb670f346bb4ab6e09047b8b5f4273f63f0a3d

**Review by anitaa1990 — APPROVED** [review]
LGTM! Nice work @hichamboushaba 🥇

**Pull request merged**